### PR TITLE
add support for variable substitution for user agent

### LINKED ISSUE #9722: Support variable substitution for user agent
### First Timers Only
We know figuring out the process for contributing to an open source project can be intimidating, so we created this issue as a way for you to learn the ropes.  (If you feel comfortable contributing to open source projects, please leave this issue for someone else.)

### What you will need to know
- Some familiarity with JavaScript

### Background
[How AMP support variable substitutions in various features](https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md)

### Motivation
When first implemented the variable substitution feature, we didn't add support to user agent, because it can be done with the user-agent header in every request. 

Some publishers however would like to collect the full user agent string as part of their analytics assessment. We are adding support to variable substitution for user agent.

Ref Issue #5050

### Step by step
- [ ] Claim this issue by adding a comment below.  Please only claim this bug if you plan on starting work in the next day or so.
- [ ] If you aren't too familiar with Git/GitHub, see the [Getting Started End-to-End Guide](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md) for [an intro to Git & GitHub,](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#intro-to-git-and-github) and [how to get a copy of the code](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#get-a-copy-of-the-amphtml-code).  You can also refer to the [Quick Start Guide](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-quick.md) for the necessary setup steps with less explanation than the End-to-End guide.
- [ ] Follow the instructions for [building AMP](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#building-amp-and-starting-a-local-server).
- [ ] Search for an existing substitute variable that we support.  (For example: [BROWSER_LANGUAGE](https://github.com/ampproject/amphtml/blob/master/spec/amp-var-substitutions.md#browser-language)).  Follow the implementing strategy of it and find all files that needs to be changed. 
- [ ] Do some manual test and sanity check to make sure everything works.
- [ ] [Create a Git branch](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#create-a-git-branch) for making your changes.
- [ ] You may also need to add unit test to `test-url-replacements.js`.
- [ ] Document the new supported variable to `amp-var-substitutions.md`.
- [ ] [Commit your changes](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#edit-files-and-commit-them) frequently.
- [ ] [Push your changes to GitHub](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#push-your-changes-to-your-github-fork).
- [ ] [Create a Pull Request](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#send-a-pull-request-ie-request-a-code-review) and request a code review from me (@zhouyx).  Mention `Fixes #9722 ` in the description.
- [ ] [Respond to your reviewer's comments](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#respond-to-pull-request-comments) (if any).

Once approved, your changes will be merged.  **⚡⚡⚡Congrats on making your first contribution to the AMP Project!⚡⚡⚡**  You'll be able to see it [live across the web soon](https://github.com/ampproject/amphtml/blob/master/contributing/release-schedule.md)!

Thanks, and we hope to see more contributions from you soon.

### Questions?

If you have questions ask in this issue or on your Pull Request (if you've created one) or see the [How to get help](https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-e2e.md#how-to-get-help) section of the Getting Started guide.

### PR DESCRIPTION
**Pull request by saqibnaveed**
Fixes #9722

Hello @zhouyx  please review this pull request. I have added variable substitution for user agent. Please let me know if there are any changes to me made.
Thanks

### CODE REVIEW

**Comment by saqibnaveed** [comment]
I signed it!

File: extensions/amp-analytics/0.1/vendors.js [hunk 1]
```diff
@@ -76,6 +76,7 @@ export const ANALYTICS_CONFIG = /** @type {!JsonObject} */ ({
       'viewer': 'VIEWER',
       'viewportHeight': 'VIEWPORT_HEIGHT',
       'viewportWidth': 'VIEWPORT_WIDTH',
+      'userAgent': 'USER_AGENT',
```
Comment (zhouyx) [inline extensions/amp-analytics/0.1/vendors.js]: The list is sorted by alphabetic order. Please follow and place `userAgent` to the correct place

File: yarn.lock [hunk 2]
```diff
@@ -48,10 +48,6 @@ acorn-jsx@^3.0.0:
   dependencies:
```
Comment (zhouyx) [inline yarn.lock]: please revert this unrelated file change

**Comment by saqibnaveed** [comment]
Okay thanks I will make these changes

**Comment by rudygalfi** [comment]
@saqibnaveed Did you get a chance to review @zhouyx's suggestions? Anything we can help with?

**Comment by saqibnaveed** [comment]
sorry I had no internet for the past 7 days. I am going to commit changes right away

**Comment by saqibnaveed** [comment]
@rudygalfi can you help me resolve conflict with yarn.lock file

I accidentally pushed it 2 commits ago.. now i am stuck

**Comment by dmorellet** [comment]
Hi @saqibnaveed, can you squash all your commits into 1 maybe ?

**Comment by lannka** [comment]
not sure what went wrong with your repo, but you probably can fix by 
1. sync your local master to upstream head
2. go to your branch and merge with master and resolve conflicts 
3. force push your branch to your origin branch

**Comment by rudygalfi** [comment]
@saqibnaveed Any updates?

**Comment by saqibnaveed** [comment]
Sorry for the delay. @lannka Thank you for your help, I think it worked. Kindly review :)

File: src/service/url-replacements-impl.js [hunk 3]
```diff
@@ -462,6 +462,12 @@ export class GlobalVariableSource extends VariableSource {
     this.set('BACKGROUND_STATE', () => {
       return viewerForDoc(this.ampdoc).isVisible() ? '0' : '1';
     });
+
+    // Returns the user agent.
+    this.set('USER_AGENT', () => {
```
Comment (zhouyx) [inline src/service/url-replacements-impl.js]: Please group user_agent to the `Device and Browser` category, around [line](https://github.com/ampproject/amphtml/blob/master/src/service/url-replacements-impl.js#L357).

File: test/functional/test-url-replacements.js [hunk 4]
```diff
@@ -950,6 +950,12 @@ describes.sandboxed('UrlReplacements', {}, () => {
         '&a=b&d=PROM&e=PAGE_LOAD_TIME');
   });
 
+  it('should replace USER_AGENT', () => {
```
Comment (zhouyx) [inline test/functional/test-url-replacements.js]: userAgent should be replaced during [page load time](https://github.com/ampproject/amphtml/blob/master/test/functional/test-url-replacements.js#L466).  Please add the test around [line](https://github.com/ampproject/amphtml/blob/master/test/functional/test-url-replacements.js#L591)

**Comment by saqibnaveed** [comment]
@zhouyx Okay thanks I will make these changes

**Comment by saqibnaveed** [comment]
@zhouyx Hello, I have made the changes you asked. Kindly review the changes

**Comment by saqibnaveed** [comment]
Locally when i run gulp test, the build is fine and my changes do not produce an error. But when travis runs tests, it says possible some dead links found in spec/amp-var-substitutions.md
I don't understand whats wrong...

**Comment by zhouyx** [comment]
@saqibnaveed The error is not related to your PR. Since you modify the `amp-var-substitutions.md` doc, we run `gulp check-links` on that file. The check then find the dead links error has been there before your change. 
@rsimha-amp I've checked the deadlink in the [doc](https://cdn.ampproject.org/c/s/example.com/page.html) `https://cdn.ampproject.org/c/s/example.com/page.html`. It seems to be a fake proxy link on purpose that we give for example. How should I fix it, and deal with these kind of dead link in future?

**Comment by rsimha-amp** [comment]
We already whitelist the base url here: https://github.com/ampproject/amphtml/blob/master/build-system/tasks/check-links.js#L130

Maybe you could expand the regex to whitelist all fake urls on that domain? Alternatively, you could enclose the link in a `<code>` block to exempt it from the link check.

**Comment by rsimha-amp** [comment]
Looking at the failure logs, the link seems to include a backtick, which seems to be confusing the link checker. The link as such is legitimate.

I'll see if I can check in a fix that prevents parsing errors due to backticks.

**Comment by zhouyx** [comment]
Thanks!
Also I think the [whitelist](https://github.com/ampproject/amphtml/blob/master/build-system/tasks/check-links.js#L131) only check the exact link `https://cdn.ampproject.org` which does not include `https://cdn.ampproject.org/c/s/example.com/page.html`. Does it makes sense to change the rule from `/https:\/\/cdn.ampproject.org(?!\/)/` to something like `/https:\/\/cdn.ampproject.org(?!\/amp-)/`?

**Comment by zhouyx** [comment]
On second thought, it's not appropriate to check the rule for some special case. I think it's maybe better to add `<code>` around

**Comment by rsimha-amp** [comment]
I found the bug. The link checker was being tripped up by empty `\`\`` blocks that were a result of the whitelisting. I'll send you a PR with a fix.

**Comment by zhouyx** [comment]
@saqibnaveed All check passed! This looks awesome! Thanks for contributing to AMP!!! 
Pull Request merged 🎉 🎉 🎉

**Comment by saqibnaveed** [comment]
Thank you for your help and merging my PR :)
